app/report_requests/process_notifications_report.py

```diff
@@ -0,0 +1,127 @@
+import csv
+from io import StringIO
+
+from flask import current_app
+from notifications_utils.s3 import (
+    s3_multipart_upload_abort,
+    s3_multipart_upload_complete,
+    s3_multipart_upload_create,
+    s3_multipart_upload_part,
+)
+
+from app.dao.notifications_dao import get_notifications_for_service
+from app.dao.report_requests_dao import dao_get_report_request_by_id
+from app.dao.service_data_retention_dao import fetch_service_data_retention_by_notification_type
+
+
+def process_report_request(service_id, report_request_id):
+    report_request = dao_get_report_request_by_id(service_id, report_request_id)
+    notification_type = report_request.parameter["notification_type"]
+    notification_status = report_request.parameter["notification_status"]
+    service_retention = fetch_service_data_retention_by_notification_type(service_id, notification_type)
+
+    page = 1
+    page_size = current_app.config.get("REPORT_REQUEST_NOTIFICATIONS_CSV_BATCH_SIZE")
+    s3_bucket = current_app.config["S3_BUCKET_REPORT_REQUESTS_DOWNLOAD"]
+    filename = f"notifications_report/{report_request_id}.csv"
+
+    init_upload_response = s3_multipart_upload_create(s3_bucket, filename)
+
+    upload_id = init_upload_response["UploadId"]
+    parts = []
+
+    csv_buffer = StringIO()
+    csv_writer = csv.writer(csv_buffer)
+
+    headers = [
+        "Recipient",
+        "Reference",
+        "Template",
+        "Type",
+        "Sent by",
+        "Sent by email",
+        "Job",
+        "Status",
+        "Time",
+        "API key name",
+    ]
+    csv_writer.writerow(headers)
+
+    try:
+        while True:
+            serialized_notifications = get_notifications_by_batch(
+                service_id=service_id,
+                template_type=report_request.parameter["notification_type"],
+                status=notification_status,
+                page=page,
+                page_size=page_size,
+                limit_days=service_retention.days_of_retention,
+            )
+
+            csv_data = convert_notifications_to_csv(serialized_notifications)
+            csv_writer.writerows(csv_data)
+
+            csv_buffer.seek(0)
+            data_bytes = csv_buffer.getvalue().encode("utf-8")
+
+            upload_response = s3_multipart_upload_part(
+                part_number=page,
+                bucket_name=s3_bucket,
+                filename=filename,
+                upload_id=upload_id,
+                data_bytes=data_bytes,
+            )
+            parts.append({"PartNumber": page, "ETag": upload_response["ETag"]})
+            page += 1
+
+            if len(serialized_notifications) < page_size:
```

Comment:
If I make `page_size` less than the number of notifications available then I get an error. I haven't dug into exactly what is failing, but we'll need to check that everything works as expected when the number of notifications is either higher than or equal to `page_size`.

Mocking `REPORT_REQUEST_NOTIFICATIONS_CSV_BATCH_SIZE` to be a low number in the tests will help with this - check out how we use the `set_config` here for example: https://github.com/alphagov/notifications-api/blob/18bd7eb06eb974627ee9f8c5c0f26e47c613779c/tests/app/dao/test_inbound_sms_dao.py#L304